packages/smart-contracts/contracts/CherryPool.sol

```diff
@@ -0,0 +1,304 @@
+     * @return 0 if successful otherwise an error code
+     */
+    function redeem(uint256 _amount)
+        public
+        isLongUtilized()
+        isShortUtilized()
+        returns (uint)
+    {
+        require(
+            _amount <= cherryDai.balanceOf(msg.sender),
+            "CherryPool::redeem request is more than current token balance"
+        );
+
+        RedeemLocalVars memory vars;
+
+        // get exchange rate from Cherrydai to Dai+fee
+        (vars.mathErr, vars.exchangeRateMantissa) = exchangeRate();
+
+        if (vars.mathErr != CherryMath.MathError.NO_ERROR) {
+            return failOpaque(Error.MATH_ERROR, FailureInfo.REDEEM_EXCHANGE_RATE_READ_FAILED, uint(vars.mathErr));
+        }
+
+        vars.redeemTokens = _amount;
+
+        // Calculate the amount of Dai to get(redeemAmount) from redeeming CherryDai(redeemTokens)
+        (vars.mathErr, vars.redeemAmount) = cherryMath.mulScalarTruncate(vars.exchangeRateMantissa, vars.redeemTokens);
+        if (vars.mathErr != CherryMath.MathError.NO_ERROR) {
+            return failOpaque(Error.MATH_ERROR, FailureInfo.REDEEM_EXCHANGE_TOKENS_CALCULATION_FAILED, uint(vars.mathErr));
+        }
+
+        /* Fail gracefully if pool has insufficient cash */
+        if (getCashPrior() < vars.redeemAmount) {
+            return fail(Error.TOKEN_INSUFFICIENT_CASH, FailureInfo.REDEEM_TRANSFER_OUT_NOT_POSSIBLE);
+        }
+
+        payout(msg.sender, vars.redeemAmount, vars.redeemTokens);
+        //require(vars.err == Error.NO_ERROR, "redeem transfer out failed");
+
+        emit RedeemCherry(msg.sender, vars.redeemAmount, vars.redeemTokens);
+
+        return uint(Error.NO_ERROR);
+
+    }
+
+    /**
+     * @dev Get available pool balance (total pool balance - total reserved balance)
+     * @return available pool balance
+     * @return 0 if successful otherwise an error code
+     */
+    function getCashPrior() internal returns (uint256) {
+        return poolBalance - (shortPoolReserved + longPoolReserved);
+    }
+
+    function payout(address _redeemer, uint256 _redeemAmount, uint256 _redeemTokens) internal returns (Error) {
+        cherryDai.burnFrom(_redeemer, _redeemTokens);
+        
+        // redeem an amount of underlying
+        require(
+            cToken.redeemUnderlying(_redeemAmount) == 0,
+            "CherryPool::payout - something went wrong"
+        );
+
+        // transfer Dai to redeemer
+        token.transfer(_redeemer, _redeemAmount);
+    }
+
+    /**
+     * @dev the rate of CherryDai redeemable for Dai.
+     * @notice Each CherryDai is convertible into the underlying asset + the fees accrued through liquidity provision.
+     * @return 0 if successful otherwise an error code
+     */
+    function exchangeRate() public returns (CherryMath.MathError, uint256) {
+        int256 rate = int256(getcTokenExchangeRate() / 1e10); //+ (poolcTokenProfit * 1e18) / int256(cherryDai.totalSupply());
+
```

Comment:
@chrismaree  I commented that line for now... it revert with invalid opcode

Reply:
@chrismaree it revert because `int256(cherryDai.totalSupply()` equal to zero in the case of the first mint()